File: trees_and_graphs/word_ladder.java
public int ladderLength(String beginWord, String endWord, List<String> wordList)
{
    // All words are same length
    int L = beginWord.length();

    // Map holding combination of words that can be formed from any given word
    // by changing one letter at a time
    Map<String, List<String>> allComboDict = new HashMap<>();

    wordList.forEach(word -> {
        for (int i = 0; i < L; i++)
        {
            // Key is the generic word
            // Value is list of words that have same intermediate generic word
            String newWord =
                word.substring(0, i) + '*' + word.substring(i + 1, L);
            
            List<String> transformations =
                allComboDict.getOrDefault(newWord, new ArrayList<>());
            
            transformations.add(word);
            allComboDict.put(newWord, transformations);
        }
    });

    // Queue for BFS
    Queue<Pair<String, Integer>> queue = new LinkedList<>();
    queue.add(new Pair(beginWord, 1));

    // visited to avoid processing same word
    Map<String, Boolean> visited = new HashMap<>();
    visited.put(beginWord, true);

    while (!queue.isEmpty())
    {
        Pair<String, Integer> node = queue.remove();
        String word = node.getKey();
        int level = node.getValue();

        for (int i = 0; i < L; i++)
        {
            // Intermediate words for current word
            String newWord =
                word.substring(0, i) + '*' + word.substring(i + 1, L);
            
            // Next states are all words that share same intermediate state
            for (String adjacentWord
                : allComboDict.getOrDefault(newWord, new ArrayList<>()))
            {
                if (adjacentWord.equals(endWord))
                {
                    return level + 1;
                }

                if (!visited.containsKey(adjacentWord))
                {
                    visited.put(adjacentWord, true);
                    queue.add(new Pair(adjacentWord, level + 1));
                }
            }
        }
    }

    return 0;
}

class Solution
{
    private int L;
    private Map<String, List<String>> allComboDict;

    Solution()
    {
        this.L = 0;

        this.allComboDict = new HashMap<>();
    }

    private int visitWordNode(
        Queue<Pair<String, Integer>> queue,
        Map<String, Integer> visited,
        Map<String, Integer> othersVisited)
    {
        int initQueueLen = queue.length();

        for (int j = 0; j < initQueueLen; j++)
        {
            Pair<String, Integer> node = queue.remove();
            String word = node.getKey();
            int level = node.getValue();

            for (int i = 0; i < this.L; i++)
            {
                String newWord =
                    word.substring(0, i) + '*' + word.substring(i + 1, L);
                
                for (String adjacentWord
                    : this.allComboDict.getOrDefault(newWord, new ArrayList<>()))
                {
                    if (othersVisited.containsKey(adjacentWord))
                    {
                        return level + othersVisited.get(adjacentWord);
                    }

                    if (!visited.containsKey(adjacentWord))
                    {
                        visited.put(adjacentWord, level + 1);
                        queue.add(new Pair(adjacentWord, level + 1));
                    }
                }
            }
        }

        return -1;
    }

    public int ladderLengthBidirBFS(
        String       beginWord,
        String       endWord,
        List<String> wordList)
    {
        if (!wordList.contains(endWord))
        {
            return 0;
        }

        this.L = beginWord.length();

        wordList.forEach( word -> {
            for (int i = 0; i < L; i++)
            {
                String newWord =
                    word.substring(0, i) + '*' + word.substring(i + 1, L);
                
                List<String> transformations =
                    this.allComboDict.getOrDefault(newWord, new ArrayList<>());
                transformations.add(word);

                this.allComboDict.put(newWord, transformations);
            }
        });

        Queue<Pair<String, Integer>> qBegin = new LinkedList<>();
        Queue<Pair<String, Integer>> qEnd = new LinkedList<>();

        qBegin.add(new Pair(beginWord, 1));
        qEnd.add(new Pair(endWord, 1));

        Map<String, Integer> visitedBegin = new HashMap<>();
        Map<String, Integer> visitedEnd = new HashMap<>();

        visitedBegin.put(beginWord, 1);
        visitedEnd.put(endWord, 1);

        int ans = -1;

        while (!qBegin.isEmpty() && !qEnd.isEmpty())
        {
            if (qBegin.size() <= qEnd.size())
            {
                ans = visitWordNode(qBegin, visitedBegin, visitedEnd);
            }
            else
            {
                ans = visitWordNode(qEnd, visitedEnd, visitedBegin);
            }

            if (ans > -1)
            {
                return ans;
            }
        }

        return 0;
    }
}

File: trees_and_graphs/word_ladder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLengthBFS = function(beginWord, endWord, wordList)
{
    const L = beginWord.length;
    
    let allComboMap = new Map();

    for (const word of wordList)
    {
        for (let i = 0; i < L; i++)
        {
            const newWord = word.substring(0, i) + "*" + word.substring(i + 1);

            if (allComboMap.has(newWord))
            {
                allComboMap.get(newWord).push(word);
            }
            else
            {
                allComboMap.set(newWord, [word]);
            }
        }
    }

    let queue = [[beginWord, 1]];
    let visited = new Set([beginWord]);

    while (queue.length)
    {
        const currQueueLength = queue.length;
        let nextQueue = [];

        for (let j = 0; j < currQueueLength; j++)
        {
            const [word, level] = queue[j];

            for (let i = 0; i < L; i++)
            {
                const newWord =
                    word.substring(0, i) + "*" + word.substring(i + 1);
                
                if (!allComboMap.has(newWord))
                {
                    continue;
                }

                for (const adjacentWord of allComboMap.get(newWord))
                {
                    if (adjacentWord == endWord)
                    {
                        return level + 1;
                    }

                    if (!visited.has(adjacentWord))
                    {
                        visited.add(adjacentWord);
                        nextQueue.push([adjacentWord, level + 1]);
                    }
                }
            }
        }

        queue = nextQueue;
    }

    return 0;
};

/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLengthBidirBFS = function(beginWord, endWord, wordList)
{
    if (!wordList.includes(endWord))
    {
        return 0;
    }

    const L = beginWord.length;

    let allComboMap = new Map();
    for (const word of wordList)
    {
        for (let i = 0; i < L; i++)
        {
            const newWord =
                word.substring(0, i) + "*" + word.substring(i + 1);

            if (!allComboMap.has(newWord))
            {
                allComboMap.set(newWord, [word]);
            }
            else
            {
                allComboMap.get(newWord).push(word);
            }
        }
    }

    let qBegin = [[beginWord, 1]];
    let qEnd = [[endWord, 1]];

    let visitedBegin = new Map();
    let visitedEnd = new Map();
    visitedBegin.set(beginWord, 1);
    visitedEnd.set(endWord, 1);

    const visitWordNode = (queue, visited, othersVisited) => {
        const currQueueLength = queue.length;

        for (let j = 0; j < currQueueLength; j++)
        {
            const [word, level] = queue.shift();

            for (let i = 0; i < L; i++)
            {
                const newWord =
                    word.substring(0, i) + "*" + word.substring(i + 1);
                
                if (!allComboMap.has(newWord))
                {
                    continue;
                }

                for (const adjacentWord of allComboMap.get(newWord))
                {
                    if (othersVisited.has(adjacentWord))
                    {
                        return level + othersVisited.get(adjacentWord);
                    }

                    if (!visited.has(adjacentWord))
                    {
                        visited.set(adjacentWord, level + 1);
                        queue.push([adjacentWord, level + 1]);
                    }
                }
            }
        }

        return -1;
    }

    let ans = -1;

    while (qBegin.length && qEnd.length)
    {
        if (qBegin.length <= qEnd.length)
        {
            ans = visitWordNode(qBegin, visitedBegin, visitedEnd);
        }
        else
        {
            ans = visitWordNode(qEnd, visitedEnd, visitedBegin);
        }

        if (ans > -1)
        {
            return ans;
        }
    }

    return 0;
};